### PR TITLE
[sw] Add :memory_unittest to //sw/device/lib/base

### PR DESCRIPTION
**Pull request by dmcardle**
Let's add a test suite for these memory functions to establish a baseline before we optimize the implementations.

Issue #13471

### CODE REVIEW

File: sw/device/lib/base/BUILD
Comment (dmcardle) [inline sw/device/lib/base/BUILD:118]: I'd like to avoid this semi-duplicate of the `:memory` target. Is there some Bazel magic that will enable us to `#define` in the `:memory_unittest` target and have it cascade down to the `:memory` dep?

Comment (cfrantz) [reply sw/device/lib/base/BUILD:121]: That's not really possible in bazel (technically it is possible via a transition to cause a re-evaluation of the build graph under the transitioning node, but that's way to complex for what you're trying to accomplish here.  The real use-case for a transition is a major configuration change, like setting the target architecture).

We have occasionally encountered this problem - the solution you have is about correct.  If the only user of `:memory_prefixed` is the unit tests, then you may want to just migrate the srcs, deps and defines into the unit test's srcs, deps and defines.

File: sw/device/lib/base/memory.h
Comment (alphan) [inline sw/device/lib/base/memory.h:37]: WDYT about slightly modifying this to become something like `OT_PREFIX_IF_NOT_RV32(name)` that could check `OT_PLATFORM_RV32`?

File: sw/device/lib/base/BUILD
Comment (alphan) [inline sw/device/lib/base/BUILD:126]: Can't we use `dual_cc_library()` for this purpose. Please see my comment about the macro before thinking about the `OT_PREFIX_MEM_FUNCTIONS` `define`.

File: sw/device/lib/base/memory.c
Comment (alphan) [inline sw/device/lib/base/memory.c:14]: nit: unfortunate but we should `undef` at the end of the header as you did.

File: sw/device/lib/base/memory_unittest.cc
Comment (alphan) [inline sw/device/lib/base/memory_unittest.cc:24]: nit: I'd just put the definitions here.

File: sw/device/lib/base/memory_unittest.cc
Comment (alphan) [inline sw/device/lib/base/memory_unittest.cc:352]: can't we use the `__builtin`s for these reference implementations? If you want to keep these, do you want to add those as well?

File: sw/device/lib/base/memory_unittest.cc
Comment (alphan) [inline sw/device/lib/base/memory_unittest.cc:32]: I like the parameterization!

**Review by alphan — APPROVED** [review]
Thanks @dmcardle!

File: sw/device/lib/base/memory.h
Comment (dmcardle) [reply sw/device/lib/base/memory.h]: It's a little verbose, but it definitely clarifies the intent at each call site. Done.

File: sw/device/lib/base/memory_unittest.cc
Comment (dmcardle) [reply sw/device/lib/base/memory_unittest.cc:352]: There are a handful reasons I opted to do this:

* When we start optimizing, I wanted to retain the dead-simple reference implementations to point to. This may be moot now that we have these parameterized unit tests.
* There isn't a corresponding builtin for each of our functions, e.g. there is no `__builtin_memrcmp`.
* I don't fully understand the semantics of the `__builtin` functions. For instance, what does  `&__builtin_memcpy` mean? I can pass that expression into the `INSTANTIATE_TEST_SUITE_P `, but inexplicably, plain `__builtin_memcpy` also works.

File: sw/device/lib/base/BUILD
Comment (dmcardle) [reply sw/device/lib/base/BUILD]: I think this comment is no longer applicable now that I took your suggestion to use the C macro `OT_PLATFORM_RV32`. Now we can use the same `:memory` target on device and host.

File: sw/device/lib/base/memory_unittest.cc
Comment (dmcardle) [reply sw/device/lib/base/memory_unittest.cc]: Done.

File: sw/device/lib/base/BUILD
Comment (alphan) [reply sw/device/lib/base/BUILD]: Sounds good (as long as we don't have a mock for this).

File: sw/device/lib/base/memory.c
Comment (alphan) [reply sw/device/lib/base/memory.c]: Yeah, but I think this comment is also not applicable if you moved this to `macros.h`, right? I added this comment just in case you decided to keep the definition in `memory.h`.

File: sw/device/lib/base/memory_unittest.cc
Comment (alphan) [reply sw/device/lib/base/memory_unittest.cc]: I thought it would be interesting to compare our implementations (both naive and optimized) to the builtins. I think it's OK to fake the ones that don't exist. 

Would you mind extending the last point? `&f` and `f` are basically the same when assigning to a function pointer, so the behavior that you described does not come as a surprise.

File: sw/device/lib/base/memory.c
Comment (alphan) [reply sw/device/lib/base/memory.c]: Sorry, just looked at the DIF and looks like you wanted to keep the definition here. SG.

File: sw/device/lib/base/memory_unittest.cc
Comment (alphan) [reply sw/device/lib/base/memory_unittest.cc]: This comment should not blocks this PR though. PLMK if you think this is ready to be merged.

Comment (dmcardle) [reply sw/device/lib/base/memory_unittest.cc]: > Would you mind extending the last point? &f and f are basically the same when assigning to a function pointer, so the behavior that you described does not come as a surprise.

Ah, I didn't realize that function lvalues implicitly converted to function pointers: https://en.cppreference.com/w/cpp/language/implicit_conversion

OK, I guess I don't a good justification for disliking the `__builtin` functions. I'm happy to delete a few reference implementations in favor of the `__builtin` functions.

**Comment by dmcardle** [comment]
This should be ready to merge now, assuming the unit tests still pass.

File: sw/device/lib/base/memory.c
Comment (dmcardle) [reply sw/device/lib/base/memory.c]: Ah, I didn't really think about moving it into macros.h.  I think if we need this kind of name mangling again, we should definitely dedupe the definition, but in the meantime I think it's nicer to define it closer to where we use it. It also gives us a nice place to hang the comment explaining the reason for name mangling.

**Comment by alphan** [comment]
> @alphan: This should be ready to merge now, assuming the unit tests still pass.

Great, thanks! I'll merge this once the CI is done.

**Comment by dmcardle** [comment]
Sweet, tests passed. Next, I'll make a PR that actually optimizes the mem functions.

**Pull request merged**
